File: dotnet/ThirtyFour.UI/Form/FormElement.cs
using System;
using System.Collections.Generic;
using System.Linq;
using OpenQA.Selenium;

namespace ThirtyFour.UI.Form
{
    /// <summary>
    /// Represents a form element with a label
    /// </summary>
    public abstract class FormElement : WrappedElement
    {
        /// <summary>
        /// Constructor
        /// </summary>
        /// <param name="element">The wrapped form element</param>
        protected FormElement(IWebElement element) : base(element) { }

        /// <summary>
        /// The value of the element
        /// </summary>
        public string Value => Element.GetAttribute("value");

        /// <summary>
        /// The label text of the element or an empty string if no associated label
        /// </summary>
        public string LabelText
        {
            get
            {
                IReadOnlyCollection<IWebElement> labels = null;
                IWebElement parentLabel;

                //attempt to find label using for=id
                string id = Element.GetAttribute("id");
                if (id != null)
                {
                    labels = Driver.FindElements(By.CssSelector(String.Format("label[for='{0}']", id)));
                }

                if (labels == null || labels.Count == 0) //attempt to find wrapped label element
                {
                    parentLabel = GetAncestorElement("label");
                }
                else
                {
                    parentLabel = labels.ElementAt(0);
                }

                return parentLabel == null ? "" : parentLabel.Text;
            }
        }
    }
}

File: java/src/main/java/com/github/danhumphrey/thirtyfour/ui/form/FormElement.java
package com.github.danhumphrey.thirtyfour.ui.form;

import java.util.ArrayList;
import java.util.List;

import org.openqa.selenium.By;
import org.openqa.selenium.WebElement;

import com.github.danhumphrey.thirtyfour.ui.WrappedElement;

/**
 * Represents a form element with a label
 */
public abstract class FormElement extends WrappedElement {

	/**
	 * Constructor
	 * @param element the wrapped form element
	 */
	public FormElement(WebElement element) {
		super(element);
	}
	
	/**
	 * Gets the label text of the element
	 * @return the text of the label associated with the element or an empty string
	 */
	public String getLabelText() {
		
		List<WebElement> labels = new ArrayList<WebElement>();
		WebElement parentLabel = null;
		
		//attempt to find label using for=id
		String id = this.element.getAttribute("id");
		if(id != null) {
			labels = this.driver.findElements(By.cssSelector(String.format("label[for='%s']",id)));
		} 
		
		if(labels.size() == 0) { //attempt to find wrapped label element
			parentLabel = this.getAncestorElement("label");
		} else {
			parentLabel = labels.get(0);
		}
		
		return parentLabel == null ? "" : parentLabel.getText();
	}
	
	/**
	 * Gets the value attribute of the element
	 * @return the attribute value
	 */
	public String getValue() {
		return this.element.getAttribute("value");
	}
}